File: app/src/main/java/com/worldventures/dreamtrips/social/ui/infopages/presenter/EnrollMemberPresenter.java
package com.worldventures.dreamtrips.social.ui.infopages.presenter;

import android.location.Location;

import com.worldventures.dreamtrips.modules.dtl.location.LocationDelegate;
import com.worldventures.dreamtrips.social.ui.infopages.StaticPageProvider;

import java.util.concurrent.TimeUnit;

import javax.inject.Inject;

import rx.Observable;

public class EnrollMemberPresenter extends AuthorizedStaticInfoPresenter {

   @Inject LocationDelegate locationDelegate;
   @Inject StaticPageProvider staticPageProvider;

   public EnrollMemberPresenter(String url) {
      super(url);
   }

   @Override
   public void load() {
      doWithLocation()
            .subscribe(location -> {
               updateUrlWithLocation(location);
               super.load();
            }, e -> super.load());
   }

   @Override
   protected void reload() {
      doWithLocation()
            .subscribe(location -> {
               updateUrlWithLocation(location);
               super.reload();
            }, e -> super.reload());
   }

   private void updateUrlWithLocation(Location location) {
      url = staticPageProvider.getEnrollWithLocation(location.getLatitude(), location.getLongitude());
   }

   // FIXME: 7/5/17 locationDelegate.requestLocationUpdate() holds reference to the context even after unSubscribe
   // Bug is known (https://github.com/mcharmas/Android-ReactiveLocation/pull/142) but the library seems to be abandoned.
   private Observable<Location> doWithLocation() {
      return Observable.merge(
            Observable.timer(2, TimeUnit.SECONDS).flatMap(unit -> Observable.error(new IllegalStateException())),
            locationDelegate.requestLocationUpdate())
            .take(1)
            .compose(bindViewToMainComposer());
   }
}

File: app/src/main/java/com/worldventures/dreamtrips/social/ui/infopages/presenter/EnrollMemberPresenter.kt
package com.worldventures.dreamtrips.social.ui.infopages.presenter

import android.location.Location

import com.worldventures.dreamtrips.social.ui.infopages.util.PermissionLocationDelegate
import com.worldventures.dreamtrips.social.ui.membership.service.analytics.EnrollMemberViewedAction
import com.worldventures.dreamtrips.social.ui.util.PermissionUIComponent

import javax.inject.Inject

class EnrollMemberPresenter : AuthorizedStaticInfoPresenter<EnrollMemberPresenter.View>() {

   @Inject lateinit var permissionLocationDelegate: PermissionLocationDelegate

   override fun initUrl() = provider.enrollMemberUrl ?: ""

   override fun takeView(view: View) {
      permissionLocationDelegate.setNeedRationalAction(view::showPermissionExplanationText)
      super.takeView(view)
      url = provider.enrollMemberUrl
   }

   public override fun load() {
      permissionLocationDelegate.apply {
         setLocationObtainedAction {
            updateUrlWithLocation(it)
            super.load()
         }

         requestPermission(true, bindView<Any>())
      }
   }

   override fun reload() {
      permissionLocationDelegate.apply {
         setLocationObtainedAction {
            updateUrlWithLocation(it)
            super.reload()
         }
         requestPermission(true, bindView<Any>())
      }
   }

   private fun updateUrlWithLocation(location: Location?) {
      location?.let { url = provider.getEnrollWithLocation(it.latitude, it.longitude) }
   }

   fun recheckPermissionAccepted(recheckAccepted: Boolean) {
      permissionLocationDelegate.recheckPermissionAccepted(recheckAccepted, bindView<Any>())
   }

   fun sendAnalyticsEnrollViewedAction() {
      analyticsInteractor.analyticsActionPipe().send(EnrollMemberViewedAction(appSessionHolder.get().get().username()))
   }

   interface View : AuthorizedStaticInfoPresenter.View, PermissionUIComponent

}
